### PR TITLE
Fix timing issue in parallel execution

### PR DESCRIPTION
**Pull request by nohwnd**
## Description

Parallel workers are able to pick up work between setting HasWork and actually associating the workload, which leads to the work being unscheduled. The test run (or test discovery) then never finishes. This was observed on a run with 99 dlls, and can be reproduced locally. To me it was on parallel runs of 12 hosts.

This fix moves HasWork to be the last item to be set under the lock, so the readers that are outside of the lock cannot read it too early before the slot is completely populated.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs [hunk 1]
```diff
@@ -152,6 +165,13 @@ private bool RunWorkInParallel()
                 slot.Work = workload.Work;
 
                 _workloads.Remove(workload);
+
+                EqtTrace.Verbose($"ParallelOperationManager.RunWorkInParallel: Adding 1 workload to slot, remaining workloads {_workloads.Count}.");
+
+                // This must be set last, every loop below looks at this property,
+                // and they can do so from a different thread. So if we mark it as HasWork before actually assigning the work
+                // we can pick up the slot, but it has no associated work yet.
+                slot.HasWork = true;
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelOperationManager.cs:174]: The fix.

**Review by Evangelink — APPROVED** [review]
Linking #4630 here to explain why we approve the change as-is.

**Pull request merged**